### ISSUE TITLE
Attempt to reload DBD/Multi.pm aborted.

### ISSUE THREAD

**Issue by james-coder** [dwright/DBD-Multi#4]
We've implemented DBD::Multi in our code exactly as your doc says. We're using mod_perl, and Apache 2.4.x, libdbd-mysql-perl 4.028, libmysqlclient18 5.5.62, (those drivers versions are because of Debian Jessie) latest master branch of this repo (as of this issue), and connecting to AWS Aurora (MySQL 5.6.10a) using DBD::Multi only during read-only queries. 

We're caching the $dbd handle through the entire lifetime of an Apache process, and everything seems to work great. However, when I let things go idle for a number of minutes, then come back to my test server and hit reload, I get an error: 

> Error: Attempt to reload DBD/Multi.pm aborted.
> Compilation failed in require at (eval 1889) line 3.

Can you please steer us in the right direction? Should we upgrade our mysql drivers? Is there a bug in DBD::Multi? Should I not be caching $dbd handles throughout an Apache child's life time?

(Also, I've never had the reload aborted error in the past prior to using DBD::Multi. (For many months with no issue.) I'm going to read through all the code and try to figure it out myself also, but DBD/DBI driver stuff is quite a lot to understand.)

Thank you, this module is great!

**Comment by dwright** [comment]
From the description of the problem, it sounds to me like your mysql server has an idle timeout configured.   Then something is trying to reconnect, but can’t.    The error “Attempt to reload” doesn’t come from DBD::Multi.   Some other module is generating that message.     

What’s not clear to me is what part DBD::Multi may be playing in this error.    

Can you please provide:

1.  A full stack backtrace at the point of the failure,

2.  A copy of the code where you are making the DBI->connect() call,  and

3.  A copy of the section of code where the error actually happens?


Thanks,
-Dan

**Comment by james-coder** [comment]
1. Stack only starts on my DBI->connect(), unfortunately. Or I'm doing it wrong. (Maybe I have to user caller() or stack trace helper library inside DBD/Multi.pm itself?
```
my $uri = "DBI:mysql:database=mysql;host=aurora-hostname-here"; #MySQL Writer
my $params;
my %dbh_options;
my $dbh;

$dbh_options{mysql_connect_timeout} = 3;

eval {
   $dbh = DBI->connect( 'dbi:Multi:', undef, undef, {
               dsns => [ # in priority order
                  # Reader (read-only) first priority for Aurora MySQL
                  10 => ['DBI:mysql:database=mysql;host=aurora-hostname-ro-here', $params->{username}, $params->{password}, \%dbh_options],
                  # Writer second priority for Aurora MySQL
                  20 => [$uri, $params->{username}, $params->{password}, \%dbh_options],
               ],
               failed_max    => 1,      # short credibility
               failed_expire => 60*60,  # long memory
               timeout       => 10,     # time out connection attempts after 10 seconds.
         });
};

if ($@) {
   // Complex internal way for getting stack-trace data, based on caller()
  my @caller;
  die "There was a problem with DBI->connect()";
}
```

2. See above.

3. I wish I knew. I have changed some things around now, removed some strange $dbh handle caching that could have been causing a problem. (And then thought about lots of edge cases where this could break our transactions. So I've carefully isolated DBD::Multi to running things that aren't in transactions, that only need a read-only query. I'm not sure if doing that will cause the DB to reconnect every time or if DBD::Multi internally caches connections already connected?)

**Comment by dwright** [comment]
>     1. Stack only starts on my DBI->connect(), unfortunately. Or I'm doing it wrong. (Maybe I have to user caller() or stack trace helper library inside DBD/Multi.pm itself?

Yes, you'll need to use a library to help you here.    Something like this might help:

```perl
   use Carp;
   $SIG{__DIE__} = sub { 
        print STDERR "----- DYING! $$ ----- \n\n";
        print STDERR Carp::longmess(@_);
        print STDERR "------ END DYIING $$ -------\n";
    };
```

**Comment by james-coder** [comment]
@dwright, I've changed my previous code above adding another part:
`$dbh_options{mysql_connect_timeout} = 3;`

Then I had the original 10 seconds from the DBD::Multi example.

I think this may have caused the odd "Attempt to reload DBD/Multi.pm aborted." problem. (If DBD::mysql was timing out early, but then DBD::Multi was expecting it to be valid for 10 seconds?)

The problem seems to have gone away. Just hoping it makes sense from a code perspective though, since it was a hard to reproduce issue.
